parcel-bundler/parcel
# Naming and targets refactor

This heavily refactors and reworks naming in Parcel 2:

* Parcel options are now pre-resolved during initialization and passed through parcel from there. A distinction between the option shape passed to the `Parcel` constructor and the shape of the resolved options is made in types.
* Namers are expected to only name the basenames of bundles within their dist directories. This basename can include path separators, and they will be appended to both dist paths and public urls for the bundles
* `distPath` is now split into `distDir` and `distEntry` on targets. This is basically the same splitting that occurred in the default namer but formalizes it, giving plugins consistent understanding of where files are written.
* `publicUrl` is now part of targets as well, and can be specified in the target entry in `package.json`, defaulting to '/'. It no longer exists as a top-level Parcel option
* `targets`, when passed to the Parcel constructor, can now be a list of strings instead of target objects (or a mix). If a string, we look up the resolved target name and include that. 
	* e.g. if you invoke `parcel build --target browser`, we find the browser target and only build that. Multiple targets are allowed in the form `parcel build --target browser --target browserModern`
* When in serving mode (server or hmr), only one target is supported, and it must be a browser target. 
* **In all development modes**, if a target is not explicitly provided via `--target` or similar, one targeting only the latest version of modern browsers is used. This is intended to reduce the amount of work we do in development (#1345)

Test Plan: 
* Existing integration tests were changed to reflect changes in `publicUrl` and continue to pass. 
* Use the cli to pass multiple targets to the server and verify it fails.
* Add a target for node and use the cli to pass this target by name to the server. Verify it fails
* Manual test of the simple example. Alter the publicUrl of a target and verify the url entered via raw runtime includes it.
